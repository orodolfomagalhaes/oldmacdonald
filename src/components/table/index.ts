export * from './table.component.ts';