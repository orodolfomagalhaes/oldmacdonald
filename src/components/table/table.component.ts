import {IPaginatedItems} from "../../interfaces/paginated-items.interface.ts";
import {FarmService} from "../../services/farm/farm.service.ts";
import {IPaginationParams} from "../../interfaces/pagination-params.interface.ts";
import {IFarm} from "../../interfaces/farm.interface.ts";
import {ICropProduction} from "../../interfaces/crop-production.interface.ts";
import {ICropType} from "../../interfaces/crop-type.interface.ts";

type PrevNext = number | null;

export class TableComponent extends HTMLElement {
  // pagination
  currentPage: number = 1;
  prev: PrevNext = null;
  pages: number = 1;
  next: PrevNext =  null;
  last: number = 0;
  pageSize: number = 10;
  // services
  farmService: FarmService;

  constructor(farmService = new FarmService()) {
    super();
    this.farmService = farmService;
  }

  async connectedCallback(): Promise<void> {
    const data = await this.getFarms();
    this.updateComponent(data);
    this.initSearchListener();
    this.initToastListener();
  }

  render(): void {
    this.innerHTML = `<div class="card p-3 table-responsive-sm mb-3">
      <table class="table table-hover table-striped" aria-label="List of farms with information about address, area, unit of measurement, and their productions.">
        <thead>
          <tr>
            <th scope="col">Farm</th>
            <th scope="col">Address</th>
            <th scope="col">Land area</th>
            <th scope="col">Unit of measure</th>
            <th scope="col">Crop production</th>
          </tr>
        </thead>
        <tbody class="table-group-divider"></tbody>
        </table>            
      </div>
      
      ${this.showPagination(this.pages, this.currentPage, this.prev, this.next)}
    `;

    this.initPaginationClickListener();
    this.initPageSelectorListener();
  }

  getPageSizeTemplate(): string {
    return `<div class="col-sm-auto">
        <div id="page-size-selector" class="d-flex align-items-center gap-3">        
          <span>Results per page</span>
          <select class="form-select w-auto" aria-label="Select the number of items to display per page">
            <option value="10" ${this.pageSize == 10 ? 'selected' : ''}>10</option>
            <option value="25" ${this.pageSize == 25 ? 'selected' : ''}>25</option>
            <option value="50" ${this.pageSize == 50 ? 'selected' : ''}>50</option>
          </select>
        </div>
      </div>
    `;
  }

  showPagination(pages: number, currentPage: number = 1, prev: PrevNext = null, next: PrevNext = null): string {
    if (pages < 2) return this.getPageSizeTemplate();
    return `<div class="pagination-container row gap-3">
      <div class="col-sm-auto">
        <nav aria-label="Table page navigation">
          <ul class="pagination mb-0">
            <li class="page-item ${prev ? '' : 'disabled'}"><a class="page-link" href="#" data-page="prev" aria-label="Go to previous page">Previous</a></li>
            ${Array.from({ length: pages }, (_, i) => `<li class='page-item ${currentPage == (i + 1) ? "active" : ""}'><a class="page-link" href="#" data-page="${i+1}">${i+1}</a></li>`).join('')}
            <li class="page-item ${next ? '' : 'disabled'}"><a class="page-link" href="#" data-page="next" aria-label="Go to next page">Next</a></li>
          </ul>
        </nav>
      </div>
      ${this.getPageSizeTemplate()}
    </div>`;
  }

  getFarms(filters?: any, pagination: IPaginationParams = {_page: this.currentPage, _per_page: this.pageSize}): Promise<IPaginatedItems> {
    return this.farmService.getAllFarms(filters, pagination);
  }

  getPaginationParams(page: string, pageSize: number = this.pageSize, currentPage = this.currentPage): IPaginationParams | undefined {
    const isInvalidOrSamePage = !page || page == String(currentPage);
    if (isInvalidOrSamePage) return;
    let selectedPage: number;
    switch (page) {
      case 'next':
        selectedPage = currentPage + 1;
        break;
      case 'prev':
        selectedPage = currentPage - 1;
        break;
      default:
        selectedPage = Number(page);
        break;
    }
    // Get data from page
    const pagination: IPaginationParams = {_page: selectedPage, _per_page: pageSize}
    return pagination;
  }

  updatePaginationComponentState(page: number, first: number, last: number): void {
    const navPagination = this.querySelector('.pagination');
    // Update currentPage value
    this.currentPage = page;
    // Remove active class
    const currentActiveItem = navPagination?.querySelector('.active');
    currentActiveItem?.classList.remove('active');
    // Add active class
    const nextActiveItem = navPagination?.querySelector(`.pagination [data-page="${page}"]`);
    nextActiveItem?.parentElement?.classList.add('active');
    // Clear prev and next disabled status
    const prevButton = navPagination?.querySelector('[data-page="prev"]');
    const nextButton = navPagination?.querySelector('[data-page="next"]');
    prevButton?.parentElement?.classList.remove('disabled');
    nextButton?.parentElement?.classList.remove('disabled');
    // Update prev status
    if (this.currentPage == first) {
      navPagination?.querySelector('[data-page="prev"]');
      prevButton?.parentElement!.classList.add('disabled');
    }
    // Update next status
    if (this.currentPage == last) {
      navPagination?.querySelector('[data-page="next"]');
      nextButton?.parentElement!.classList.add('disabled');
    }
  }

  updateTableData(data: IFarm[]): any {
    const tbody = this.querySelector('table tbody');
    tbody!.innerHTML = '';

    if (!data.length) {
      const tr = document.createElement('tr');
      tr.innerHTML = `<td colspan="5" class="text-center">No records found</td>`;
      tbody!.appendChild(tr);
      return;
    }

    data.forEach((item: any) => {
      const tr = document.createElement('tr');
      tr.style.cursor = 'pointer';

      tr.addEventListener('click', (() => {
        const event = new CustomEvent('show-farm-detail', {detail: item, bubbles: true});
        document.dispatchEvent(event);
      }));

      const tdName = document.createElement('td');
      tdName.innerText = item.farmName;
      tr.appendChild(tdName);

      const tdAddress = document.createElement('td');
      tdAddress.innerText = item.address;
      tr.appendChild(tdAddress);

      const tdArea = document.createElement('td');
      tdArea.innerText = item.landArea;
      tr.appendChild(tdArea);

      const tdMeasure = document.createElement('td');
      tdMeasure.innerText = item.unitOfMeasure;
      tr.appendChild(tdMeasure);

      const tdCrops = document.createElement('td');
      tdCrops.innerText = this.mapCropProductions(item.cropProductions);
      tr.appendChild(tdCrops);

      tbody!.appendChild(tr);
    });
  }

  mapCropProductions(items: ICropProduction[]) {
    return items.map((item: ICropProduction) => this.getCropTypesMap()[item.cropTypeId]).join(', ');
  }

  getCropTypes(): ICropType[] {
    const app: any = document.querySelector('app-component');
    return app ? app.getCropTypeList() : [];
  }

  getCropTypesMap() {
    return Object.fromEntries(this.getCropTypes().map((item: any) => [item.id, item.name]));
  }

  searchListenerEvent: any;
  initSearchListener(): void {
    this.searchListenerEvent = async (event: Event) => {
      const { detail } = event as CustomEvent;
      const filter = detail ? { farmName: detail } : null;
      const data = await this.getFarms(filter);
      this.updateComponent(data);
    }
    document.addEventListener('search-event', this.searchListenerEvent);
  }

  initPageSelectorListener(): void {
    const select: any = this.querySelector('#page-size-selector select');
    select?.addEventListener('change', async () => {
      // Update currentPage and pageSize
      this.pageSize = Number(select.value);

      // Clear search filter
      const searchElement: any = document.querySelector('#search-input');
      const filter = searchElement?.value ? { farmName: searchElement.value } : null;

      this.currentPage = 1; // reset pagination
      const data = await this.getFarms(filter);
      this.updateComponent(data);
    });
  }

  initPaginationClickListener(): void {
    // Create a listener to pagination
    const pagination = this.querySelector('.pagination');
    if (!pagination) return;
    pagination.addEventListener('click', async (event: any) => {
      // Get pagination params
      const paginationParams = this.getPaginationParams(event.target.dataset.page);
      if (!paginationParams) return;
      // Make a new request with params
      const filters = undefined;
      const {first, last, data} = await this.getFarms(filters, paginationParams);
      this.updatePaginationComponentState(paginationParams._page!, first, last);
      this.updateTableData(data);
    });
  }

  updateComponentToastEvent: any;
  initToastListener(): void {
    this.updateComponentToastEvent = async () => {
      this.currentPage = 1 // reset
      const data = await this.getFarms();
      this.updateComponent(data);
    }
    document.addEventListener('show-toast', this.updateComponentToastEvent);
  }

  async updateComponent(getFarmsResponse: IPaginatedItems): Promise<void> {
    this.prev = getFarmsResponse.prev;
    this.next = getFarmsResponse.next;
    this.pages = getFarmsResponse.pages;
    this.last = getFarmsResponse.last;

    this.render();
    this.updateTableData(getFarmsResponse.data);
  }
}

customElements.define('table-component', TableComponent);