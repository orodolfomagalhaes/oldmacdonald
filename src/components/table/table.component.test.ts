import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { TableComponent } from './table.component';
import {IPaginatedItems} from "../../interfaces/paginated-items.interface.ts";
import {FarmService} from "../../services/farm/farm.service.ts";

const mockFarmData: IPaginatedItems = {
  data: [
    { id: '789', farmName: 'Fazenda Boa Vista', address: 'Test Address', landArea: 100, unitOfMeasure: 'hectares',
      cropProductions: [
        {id: 'cp001',cropTypeId: "ct001",isIrrigated: true,isInsured: false,farmId: '789'},
        {id: 'cp002',cropTypeId: "ct002",isIrrigated: true,isInsured: false,farmId: '789'}
      ]
    },
    { id: '123', farmName: 'Fazenda Nova Esperança', address: 'Test Address', landArea: 100, unitOfMeasure: 'hectares',
      cropProductions: [
        {id: 'cp001',cropTypeId: "ct001",isIrrigated: true,isInsured: false,farmId: '123'},
        {id: 'cp002',cropTypeId: "ct002",isIrrigated: true,isInsured: false,farmId: '123'}
      ]
    }
  ],
  first: 1,
  items: 2,
  last: 1,
  next: null,
  pages: 1,
  prev: null
};
describe('TableComponent', () => {
  let component: TableComponent;
  globalThis.fetch = vi.fn();

  beforeEach(() => {
    component = new TableComponent();
    component.getFarms = vi.fn().mockResolvedValue(mockFarmData);
    document.body.appendChild(component);
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.resetAllMocks();

    // disable events for tests to enable when needed
    document.removeEventListener('show-toast', component.updateComponentToastEvent);
    document.removeEventListener('search-event', component.searchListenerEvent);
  });

  it('should create component', async () => {
    // disable show events for all tests to enable when needed
    document.removeEventListener('show-toast', component.updateComponentToastEvent);

    const noFarmDataMock = {data: [],first: 1,items: 0,last: 1,next: null,pages: 1,prev: null}
    component.getFarms = vi.fn().mockResolvedValue(noFarmDataMock);
    const updateComponentSpy = vi.spyOn(component, 'updateComponent');
    const getFarmsSpy = vi.spyOn(component, 'getFarms');
    await component.connectedCallback();
    const tbody = component.querySelector('table tbody');

    expect(component).toBeTruthy();
    expect(getFarmsSpy).toHaveBeenCalledOnce();
    expect(updateComponentSpy).toHaveBeenNthCalledWith(1, noFarmDataMock);
    expect(tbody!.innerHTML).toContain('<td colspan="5" class="text-center">No records found</td>');
    expect(component).toMatchSnapshot();

  });

  it('should render table with correct structure', () => {
    // disable events for tests to enable when needed
    document.removeEventListener('show-toast', component.updateComponentToastEvent);

    const paginationMock = document.createElement('nav');
    vi.spyOn(component, 'querySelector').mockReturnValue(paginationMock);
    vi.spyOn(component, 'showPagination').mockReturnValue('<div>Pagination Mock</div>');
    component.render();

    expect(component.innerHTML).toContain('table-responsive-sm');
    expect(component.innerHTML).toContain('aria-label="List of farms');
    expect(component.innerHTML).toContain('<thead>');
    expect(component.innerHTML).toContain('<tbody class="table-group-divider">');
  });

  it('should add data to the table when available', async () => {
    // disable events for tests to enable when needed
    document.removeEventListener('search-event', component.searchListenerEvent);
    document.removeEventListener('show-toast', component.updateComponentToastEvent);

    await component.connectedCallback();
    const rows = component.querySelectorAll('table tbody tr');

    expect(rows.length).toBe(2);
    expect(rows[0].querySelectorAll('td')[0].innerText).toEqual('Fazenda Boa Vista');
    expect(rows[1].querySelectorAll('td')[0].innerText).toBe('Fazenda Nova Esperança');

  });

  it('should dispatch an event when a table row is clicked', async () => {
    // disable events for tests to enable when needed
    document.removeEventListener('search-event', component.searchListenerEvent);
    document.removeEventListener('show-toast', component.updateComponentToastEvent);

    const eventListener = vi.fn();
    document.addEventListener('show-farm-detail', eventListener);
    await component.connectedCallback();
    const row: any = component.querySelector('table tbody tr')!;
    row.click();

    const capturedEvent = eventListener.mock.calls[0][0];

    expect(eventListener).toHaveBeenCalledTimes(1);
    expect(capturedEvent.type).toBe('show-farm-detail');
    expect(capturedEvent.detail).toEqual(mockFarmData.data[0]);

  });

  it('should update table when show-toast event is captured', async () => {
    const updateComponentSpy = vi.spyOn(component, 'updateComponent');
    const customEvent = new CustomEvent('show-toast', {
      detail: { message: 'Data updated', type: 'success' },
      bubbles: true,
      cancelable: true,
    });
    document.dispatchEvent(customEvent);
    await Promise.resolve();

    expect(component.currentPage).toBe(1);
    expect(component.pages).toBe(1);
    expect(updateComponentSpy).toHaveBeenCalledTimes(1);
  });

  it('should correctly map crop productions to names', () => {
    const cropProductions = [
      {
        id: 'cp001',
        cropTypeId: "ct001",
        isIrrigated: true,
        isInsured: false,
        farmId: '789'
      },
      {
        id: 'cp002',
        cropTypeId: "ct002",
        isIrrigated: true,
        isInsured: false,
        farmId: '789'
      }
    ];
    component.getCropTypesMap = vi.fn().mockReturnValue({
      "ct001": "Soja",
      "ct002": "Milho"
    })
    const result = component.mapCropProductions(cropProductions);

    expect(result).toBe('Soja, Milho');
  });

  it('should init pagination click listener', async () => {
    document.removeEventListener('search-event', component.searchListenerEvent);
    const paginationListenerSpy = vi.spyOn(component, 'initPaginationClickListener');
    await component.connectedCallback();

    expect(paginationListenerSpy).toHaveBeenCalledTimes(1);
  });

  it('should handle "next" page navigation', () => {
    component.currentPage = 1;
    const result = component.getPaginationParams('next');
    expect(result).toEqual({ _page: 2, _per_page: 10 });
  });

  it('should handle "prev" page navigation', () => {
    component.currentPage = 2;
    const result = component.getPaginationParams('prev');
    expect(result).toEqual({ _page: 1, _per_page: 10 });
  });

  it('should handle numeric page navigation', () => {
    const result = component.getPaginationParams('3');
    expect(result).toEqual({ _page: 3, _per_page: 10 });
  });

  it('should return undefined for invalid or same page', () => {
    expect(component.getPaginationParams('')).toBeUndefined();
    expect(component.getPaginationParams('1')).toBeUndefined();
  });

  it('should to next page', async () => {
    document.removeEventListener('search-event', component.searchListenerEvent);
    const mockedPageOne: IPaginatedItems = {
      data: [
        { id: '789', farmName: 'Fazenda Boa Vista', address: 'Test Address', landArea: 100, unitOfMeasure: 'hectares',
          cropProductions: [
            {id: 'cp001',cropTypeId: "ct001",isIrrigated: true,isInsured: false,farmId: '789'},
            {id: 'cp002',cropTypeId: "ct002",isIrrigated: true,isInsured: false,farmId: '789'}
          ]
        }
      ],
      first: 1,
      items: 2,
      last: 2,
      next: 2,
      pages: 2,
      prev: null
    };
    const mockedPageTwo: IPaginatedItems = {
      data: [
        { id: '123', farmName: 'Fazenda Nova Esperança', address: 'Test Address', landArea: 100, unitOfMeasure: 'hectares',
          cropProductions: [
            {id: 'cp001',cropTypeId: "ct001",isIrrigated: true,isInsured: false,farmId: '123'},
            {id: 'cp002',cropTypeId: "ct002",isIrrigated: true,isInsured: false,farmId: '123'}
          ]
        }
      ],
      first: 1,
      items: 2,
      last: 2,
      next: null,
      pages: 2,
      prev: 1
    };
    component.getFarms = vi.fn().mockResolvedValue(mockedPageOne);
    component.pageSize = 1;
    await component.connectedCallback();

    const getPaginationParamsSpy = vi.spyOn(component, 'getPaginationParams');
    const updatePaginationComponentStateSpy = vi.spyOn(component, 'updatePaginationComponentState');
    const updateTableDataSpy = vi.spyOn(component, 'updateTableData');
    component.getFarms = vi.fn().mockResolvedValue(mockedPageTwo);

    const pagination = document.querySelector('.pagination');
    const pageTwo: any = pagination?.querySelector('[data-page="2"]');
    pageTwo?.click();

    expect(getPaginationParamsSpy).toHaveBeenCalledTimes(1);
    expect(getPaginationParamsSpy).toHaveBeenCalledWith('2');
    expect(getPaginationParamsSpy).toReturnWith({_page: 2, _per_page: component.pageSize });
    await expect(component.getFarms).toHaveBeenCalledWith(undefined, {_page: 2, _per_page: component.pageSize });

    expect(updatePaginationComponentStateSpy).toHaveBeenCalledTimes(1);
    expect(updateTableDataSpy).toHaveBeenCalledTimes(1);
    expect(updateTableDataSpy).toHaveBeenCalledWith(mockedPageTwo.data);
    expect(component.currentPage).toBe(2);

    const pageOne: any = pagination?.querySelector('[data-page="1"]');
    pageOne.click();
    expect(getPaginationParamsSpy).toHaveBeenCalledTimes(2);
    expect(getPaginationParamsSpy).toHaveBeenCalledWith('1');
    expect(getPaginationParamsSpy).toReturnWith({_page: 1, _per_page: component.pageSize });
    await Promise.resolve()
    expect(component.currentPage).toBe(1);

    pageOne.click();
    expect(getPaginationParamsSpy).toHaveNthReturnedWith(3, undefined);
  });

  it('should init page selector listener', async () => {
    document.removeEventListener('search-event', component.searchListenerEvent);
    const pageSelectorListenerSpy = vi.spyOn(component, 'initPageSelectorListener');
    await component.connectedCallback();

    expect(pageSelectorListenerSpy).toHaveBeenCalledTimes(1);
  });

  it('should call farmService getAllFarms', () => {
    class MockedFarmService extends FarmService {
      getAllFarms: any = () => {
        vi.fn().mockResolvedValue(mockFarmData)
      }
    }
    const mockedFarmService = new MockedFarmService()
    const component = new TableComponent(mockedFarmService);
    const getAllFarmsSpy = vi.spyOn(mockedFarmService, 'getAllFarms');
    component.getFarms();

    expect(getAllFarmsSpy).toHaveBeenCalledTimes(1);
  });

  it('should handle page size selection', async () => {
    await component.connectedCallback();
    let pageSizeSelector: any = component.querySelector('#page-size-selector select');
    expect(pageSizeSelector.value).toBe('10');

    const updateComponentSpy = vi.spyOn(component, 'updateComponent');
    pageSizeSelector.value = '25';
    let changeEvent = new Event('change');
    pageSizeSelector?.dispatchEvent(changeEvent);
    expect(component.pageSize).toBe(25);

    await Promise.resolve()
    expect(updateComponentSpy).toHaveBeenCalledTimes(1);

    pageSizeSelector.value = '50';
    pageSizeSelector?.dispatchEvent(changeEvent);
    expect(component.pageSize).toBe(50);

    await Promise.resolve()
    expect(updateComponentSpy).toHaveBeenCalledTimes(2);

  });

  it('should handle search listener', async () => {
    const searchEventListener = vi.fn();
    document.addEventListener('search-event', searchEventListener);
    const initSearchListenerSpy = vi.spyOn(component, 'initSearchListener');
    await component.connectedCallback();
    const updateComponentSpy = vi.spyOn(component, 'updateComponent');
    let searchEvent = new CustomEvent('search-event', { detail: 'Farm name', bubbles: true });
    component.dispatchEvent(searchEvent);

    expect(initSearchListenerSpy).toHaveBeenCalledTimes(1);
    expect(searchEventListener).toHaveBeenCalledTimes(1);
    await Promise.resolve();
    expect(updateComponentSpy).toHaveBeenCalled();

    searchEvent = new CustomEvent('search-event', { detail: '', bubbles: true });
    component.dispatchEvent(searchEvent);
    expect(searchEventListener).toHaveBeenCalledTimes(2);
  });

  it('should get app CropTypes list', () => {
    const app: any = document.createElement('app-component');
    app.getCropTypeList = vi.fn().mockReturnValue([]);
    document.body.appendChild(app);
    const getCropTypes = vi.spyOn(component, 'getCropTypes');
    component.getCropTypes();
    expect(getCropTypes).toHaveBeenCalled();
    expect(getCropTypes).toReturnWith([]);
  });

});