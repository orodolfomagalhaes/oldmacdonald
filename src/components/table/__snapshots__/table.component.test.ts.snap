// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableComponent > should create component 1`] = `
<table-component>
  <div
    class="card p-3 table-responsive-sm mb-3"
  >
    
      
    <table
      aria-label="List of farms with information about address, area, unit of measurement, and their productions."
      class="table table-hover table-striped"
    >
      
        
      <thead>
        
          
        <tr>
          
            
          <th
            scope="col"
          >
            Farm
          </th>
          
            
          <th
            scope="col"
          >
            Address
          </th>
          
            
          <th
            scope="col"
          >
            Land area
          </th>
          
            
          <th
            scope="col"
          >
            Unit of measure
          </th>
          
            
          <th
            scope="col"
          >
            Crop production
          </th>
          
          
        </tr>
        
        
      </thead>
      
        
      <tbody
        class="table-group-divider"
      >
        <tr>
          <td
            class="text-center"
            colspan="5"
          >
            No records found
          </td>
        </tr>
      </tbody>
      
        
    </table>
                
      
  </div>
  
      
      
  <div
    class="col-sm-auto"
  >
    
        
    <div
      class="d-flex align-items-center gap-3"
      id="page-size-selector"
    >
              
          
      <span>
        Results per page
      </span>
      
          
      <select
        aria-label="Select the number of items to display per page"
        class="form-select w-auto"
      >
        
            
        <option
          selected=""
          value="10"
        >
          10
        </option>
        
            
        <option
          value="25"
        >
          25
        </option>
        
            
        <option
          value="50"
        >
          50
        </option>
        
          
      </select>
      
        
    </div>
    
      
  </div>
  
    
    
</table-component>
`;
